src/app/editor/preview/layer-settings/settings-html-mapper/settings-html-mapper.component.ts

```diff
@@ -0,0 +1,176 @@
+      </mat-expansion-panel-header>
+
+      <mat-form-field class="large" [title]="selectedHtmlElementDescription">
+        <input
+          type="text"
+          matInput
+          placeholder="Attach an HTML element"
+          [ngModel]="currentLayer?.web?.tag?.name"
+          (ngModelChange)="ngModelChanged($event)"
+          [matAutocomplete]="autoGroup"
+          (blur)="updateLayerTagNameOnBlur()"
+        />
+        <mat-autocomplete
+          #autoGroup="matAutocomplete"
+          [autoActiveFirstOption]="false"
+          (optionSelected)="optionSelected($event)"
+        >
+          <mat-optgroup
+            *ngFor="let group of htmlMapperGroupAutocomplete"
+            [label]="group.type"
+          >
+            <mat-option *ngFor="let item of group.elements" [value]="item.name">
+              {{ item.name }}
+            </mat-option>
+          </mat-optgroup>
+        </mat-autocomplete>
+      </mat-form-field>
+    </mat-expansion-panel>
+  `,
+  styles: [
+    `
+      :host {
+        text-align: left;
+      }
+      mat-form-field {
+        padding: 14px;
+      }
+      mat-hint {
+        height: 130px;
+        padding: 14px;
+      }
+    `
+  ]
+})
+export class SettingsHtmlMapperComponent implements OnInit {
+  stateGroups: HtmlElementGroup[];
+
+  currentLayer: SketchMSLayer;
+
+  DEFAULT_ELEMENT = 'div';
+
+  selectedHtmlElementName = '';
+  selectedHtmlElementDescription = '';
+
+  htmlMapperGroupAutocomplete: HtmlElementGroup[];
+
+  @ViewChild(MatAutocompleteTrigger) autocompletePanel: MatAutocompleteTrigger;
+
+  constructor(
+    private readonly store: Store,
+    private readonly htmlMapper: SettingsHtmlMapperService
+  ) {}
+
+  ngOnInit() {
+    this.stateGroups = this.htmlMapper.getHtmlElements();
+
+    this.store.select(UiState.currentLayer).subscribe(currentLayer => {
+      if (currentLayer) {
+        this.autocompletePanel.closePanel();
+
+        this.currentLayer = currentLayer;
+        const value = this.currentLayer.web?.tag?.name;
+        this.selectedHtmlElementDescription = this.getSelectedHtmlElementDescription(
+          value
+        );
+      }
+    });
+  }
+
+  ngModelChanged(value: string) {
+    if (value) {
+      this.htmlMapperGroupAutocomplete = this.filterGroup(value);
+      const desc = this.getSelectedHtmlElementDescription(value);
+      if (desc) {
+        this.selectedHtmlElementDescription = desc;
+      } else {
+        // element not found in the list, update description accordingly
+        if (value.includes('-')) {
+          this.selectedHtmlElementDescription = 'Custom element';
+        } else {
+          this.selectedHtmlElementDescription = 'Unsupported element';
+        }
+      }
+    } else {
+      // reset state
+      this.htmlMapperGroupAutocomplete = this.stateGroups;
+      this.selectedHtmlElementDescription = '';
+      this.autocompletePanel.closePanel();
+      this.store.dispatch(new LayerHtmlCustomTag(null));
+    }
+  }
```

Comment:
Do you think we should refactor this to separate functions with correct names? For example, `resetstate` can be a function for line 129-133. It think, by doing this we have a better understanding of the goal of this code.